core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -775,6 +775,13 @@ sealed trait ZIO[-R, +E, +A] extends Serializable { self =>
   final def on(ec: ExecutionContext): ZIO[R, E, A] =
     self.lock(Executor.fromExecutionContext(Int.MaxValue)(ec))
 
+  /**
+   * Returns an effect that will only be executed once, even if it is evalauted
+   * multiple times.
+   */
+  final def once: ZIO[Any, Nothing, ZIO[R, E, Unit]] =
+    Ref.make(true).map(ref => self.whenM(ref.getAndSet(false)))
```

Comment:
The semantics here currently are "execute at most once", since errors / interruption will leave the flag set but `self` was not necessarily performed. Maybe we should use `bracketExit` here?

Also, it'd be nice to make this return the value `A` with a fallback value/effect when the ref is already set. This is the prevalent pattern in ZStream.

Reply:
I'm not sure what the right semantics are here. My original intention was "guarantee that this effect will only begin execution once" so you could pass the effect around and get the same semantics as if you only evaluated it once (and it potentially was interrupted or failed). Resetting the flag raises the possibility that the "work" will be done multiple times if there are failures or the effect is interrupted after doing significant work. We could definitely generalize by allowing the caller to provide a `Exit[E, A] => Boolean` function and a default value, thought it does complicate the method signature.

Reply:
`onceWith/M` sounds like a great variant :-)